scipy/sparse/linalg/_norm.py

```diff
@@ -177,6 +177,11 @@ def norm(x, ord=None, axis=None):
             except TypeError as e:
                 raise ValueError('Invalid norm order for vectors.') from e
             M = np.power(abs(x).power(ord).sum(axis=a), 1 / ord)
-        return M.A.ravel()
+        if hasattr(M, 'toarray'):
+            return M.toarray().ravel()
+        elif hasattr(M, 'A'):
+            return M.A.ravel()
+        else:
+            return M.ravel()
```

Comment:
I'm confused by these branches... both sparse matrix and sparse array objects have a `toarray` method, so can't the `elif` branch be dropped?

Reply:
I couldn't figure out a simpler way to handle all the cases.  For instance, summing a sparse matrix along an axis yields a dense matrix (with `.A` but no `.toarray`).

Reply:
Ah right - the sum results in an `np.matrix` object, not a sparse object, which has `.A` but not `.toarray` - I understand now :+1: